File: padlock/src/main/java/com/pyamsoft/padlock/list/LockInfoDialog.java
/*
 * Copyright 2017 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.padlock.list;

import android.app.Dialog;
import android.os.Bundle;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.support.v4.view.ViewCompat;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Toast;
import com.mikepenz.fastadapter.commons.adapters.FastItemAdapter;
import com.pyamsoft.padlock.Injector;
import com.pyamsoft.padlock.PadLock;
import com.pyamsoft.padlock.R;
import com.pyamsoft.padlock.databinding.DialogLockInfoBinding;
import com.pyamsoft.padlock.loader.AppIconLoader;
import com.pyamsoft.padlock.model.ActivityEntry;
import com.pyamsoft.padlock.model.AppEntry;
import com.pyamsoft.pydroid.loader.ImageLoader;
import com.pyamsoft.pydroid.loader.LoaderHelper;
import com.pyamsoft.pydroid.loader.loaded.Loaded;
import com.pyamsoft.pydroid.ui.util.DialogUtil;
import com.pyamsoft.pydroid.util.AppUtil;
import javax.inject.Inject;
import timber.log.Timber;

public class LockInfoDialog extends DialogFragment {

  @NonNull public static final String TAG = "LockInfoDialog";
  @NonNull private static final String ARG_APP_PACKAGE_NAME = "app_packagename";
  @NonNull private static final String ARG_APP_NAME = "app_name";
  @NonNull private static final String ARG_APP_SYSTEM = "app_system";
  @SuppressWarnings("WeakerAccess") @Inject LockInfoPresenter presenter;
  @SuppressWarnings("WeakerAccess") FastItemAdapter<LockInfoItem> fastItemAdapter;
  @SuppressWarnings("WeakerAccess") DialogLockInfoBinding binding;
  String appPackageName;
  boolean appIsSystem;
  boolean listDoneRefreshing;
  private String appName;
  private DividerItemDecoration dividerDecoration;
  private FilterListDelegate filterListDelegate;
  @NonNull private Loaded appIcon = LoaderHelper.empty();

  @CheckResult @NonNull public static LockInfoDialog newInstance(@NonNull AppEntry appEntry) {
    final LockInfoDialog fragment = new LockInfoDialog();
    final Bundle args = new Bundle();

    args.putString(ARG_APP_PACKAGE_NAME, appEntry.packageName());
    args.putString(ARG_APP_NAME, appEntry.name());
    args.putBoolean(ARG_APP_SYSTEM, appEntry.system());

    fragment.setArguments(args);
    return fragment;
  }

  @NonNull @Override public Dialog onCreateDialog(Bundle savedInstanceState) {
    final Dialog dialog = super.onCreateDialog(savedInstanceState);
    dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
    return dialog;
  }

  @Override public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    listDoneRefreshing = false;

    appPackageName = getArguments().getString(ARG_APP_PACKAGE_NAME, null);
    appName = getArguments().getString(ARG_APP_NAME, null);
    appIsSystem = getArguments().getBoolean(ARG_APP_SYSTEM, false);

    if (appPackageName == null || appName == null) {
      throw new NullPointerException("App information is NULL");
    }

    Injector.get().provideComponent().inject(this);
  }

  @Nullable @Override
  public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
      @Nullable Bundle savedInstanceState) {
    filterListDelegate = new FilterListDelegate();
    fastItemAdapter = new FastItemAdapter<>();
    binding = DialogLockInfoBinding.inflate(inflater, container, false);
    return binding.getRoot();
  }

  @Override public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);
    setupToolbar();
    binding.lockInfoPackageName.setText(appPackageName);
    binding.lockInfoSystem.setText((appIsSystem ? "YES" : "NO"));
    filterListDelegate.onViewCreated(fastItemAdapter);
    setupSwipeRefresh();
    setupRecyclerView();
  }

  private void setupToolbar() {
    ViewCompat.setElevation(binding.lockInfoToolbar, AppUtil.convertToDP(getContext(), 4));
    binding.lockInfoToolbar.setTitle(appName);
    binding.lockInfoToolbar.setNavigationOnClickListener(v -> dismiss());
    binding.lockInfoToolbar.inflateMenu(R.menu.search_menu);
    filterListDelegate.onPrepareOptionsMenu(binding.lockInfoToolbar.getMenu(), fastItemAdapter);
  }

  private void setupSwipeRefresh() {
    binding.lockInfoSwipeRefresh.setColorSchemeResources(R.color.blue500, R.color.amber700,
        R.color.blue700, R.color.amber500);
    binding.lockInfoSwipeRefresh.setOnRefreshListener(() -> {
      Timber.d("onRefresh");
      refreshList(true);
    });
  }

  private void setupRecyclerView() {
    dividerDecoration = new DividerItemDecoration(getActivity(), DividerItemDecoration.VERTICAL);

    fastItemAdapter.getItemFilter().withFilterPredicate((item, query) -> {
      final String queryString = String.valueOf(query).toLowerCase().trim();
      return item.filterAgainst(queryString);
    });

    LinearLayoutManager manager = new LinearLayoutManager(getContext());
    manager.setItemPrefetchEnabled(true);
    manager.setInitialPrefetchItemCount(3);
    binding.lockInfoRecycler.setLayoutManager(manager);
    binding.lockInfoRecycler.setClipToPadding(false);
    binding.lockInfoRecycler.setHasFixedSize(false);
    binding.lockInfoRecycler.addItemDecoration(dividerDecoration);
    binding.lockInfoRecycler.setAdapter(fastItemAdapter);
  }

  @Override public void onDestroyView() {
    super.onDestroyView();
    filterListDelegate.onDestroyView();

    binding.lockInfoRecycler.removeItemDecoration(dividerDecoration);
    binding.lockInfoRecycler.setOnClickListener(null);
    binding.lockInfoRecycler.setLayoutManager(null);
    binding.lockInfoRecycler.setAdapter(null);
    binding.unbind();
  }

  @Override public void onDestroy() {
    super.onDestroy();
    presenter.destroy();
    PadLock.getRefWatcher(this).watch(this);
  }

  @Override public void onStart() {
    super.onStart();

    appIcon = LoaderHelper.unload(appIcon);
    appIcon = ImageLoader.fromLoader(AppIconLoader.forPackageName(appPackageName))
        .into(binding.lockInfoIcon);

    if (!listDoneRefreshing) {
      refreshList(false);
    }
  }

  @Override public void onStop() {
    super.onStop();
    presenter.stop();
    appIcon = LoaderHelper.unload(appIcon);
  }

  @Override public void onResume() {
    super.onResume();
    // The dialog is super small for some reason. We have to set the size manually, in onResume
    final Window window = getDialog().getWindow();
    if (window != null) {
      window.setLayout(WindowManager.LayoutParams.MATCH_PARENT,
          WindowManager.LayoutParams.WRAP_CONTENT);
    }
  }

  void refreshList(boolean force) {
    presenter.populateList(appPackageName, force, new LockInfoPresenter.PopulateListCallback() {

      private void setRefreshing(boolean refresh) {
        // In case the configuration changes, we do the animation again
        binding.lockInfoSwipeRefresh.post(() -> {
          if (binding != null) {
            if (binding.lockInfoSwipeRefresh != null) {
              binding.lockInfoSwipeRefresh.setRefreshing(refresh);
            }
          }
        });
      }

      @Override public void onListPopulateBegin() {
        listDoneRefreshing = false;
        setRefreshing(true);
        fastItemAdapter.clear();
      }

      @Override public void onEntryAddedToList(@NonNull ActivityEntry entry) {
        fastItemAdapter.add(new LockInfoItem(appPackageName, appIsSystem, entry));
      }

      @Override public void onListPopulated() {
        listDoneRefreshing = true;
        setRefreshing(false);

        if (fastItemAdapter.getAdapterItemCount() > 0) {
          Timber.d("Refresh finished");
          presenter.showOnBoarding(new LockInfoPresenter.OnBoardingCallback() {
            @Override public void onShowOnboarding() {
              Timber.d("Show onboarding");
              // TODO
            }

            @Override public void onOnboardingComplete() {
              // TODO
            }
          });
        } else {
          Toast.makeText(getContext(), "Error while loading list. Please try again.",
              Toast.LENGTH_SHORT).show();
        }
      }

      @Override public void onListPopulateError() {
        DialogUtil.guaranteeSingleDialogFragment(getActivity(), new ErrorDialog(), "error");
      }
    });
  }
}

File: padlock/src/main/java/com/pyamsoft/padlock/list/LockInfoDialog.kt
/*
 * Copyright 2019 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.pyamsoft.padlock.list

import android.os.Bundle
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import androidx.annotation.CheckResult
import com.pyamsoft.padlock.Injector
import com.pyamsoft.padlock.PadLockComponent
import com.pyamsoft.padlock.R
import com.pyamsoft.padlock.list.info.LockInfoViewModel
import com.pyamsoft.padlock.loader.AppIconLoader
import com.pyamsoft.padlock.model.list.ActivityEntry
import com.pyamsoft.padlock.model.list.AppEntry
import com.pyamsoft.padlock.model.list.ListDiffProvider
import com.pyamsoft.pydroid.core.singleDisposable
import com.pyamsoft.pydroid.core.tryDispose
import com.pyamsoft.pydroid.ui.app.fragment.ToolbarDialog
import com.pyamsoft.pydroid.ui.app.fragment.requireArguments
import com.pyamsoft.pydroid.ui.util.show
import javax.inject.Inject

class LockInfoDialog : ToolbarDialog() {

  @field:Inject internal lateinit var appIconLoader: AppIconLoader
  @field:Inject internal lateinit var viewModel: LockInfoViewModel
  @field:Inject internal lateinit var lockView: LockInfoView

  private var databaseChangeDisposable by singleDisposable()
  private var lockEventDisposable by singleDisposable()
  private var populateDisposable by singleDisposable()

  override fun onCreateView(
    inflater: LayoutInflater,
    container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View? {
    val appPackageName = requireArguments().getString(ARG_APP_PACKAGE_NAME, "")
    val appName = requireArguments().getString(ARG_APP_NAME, "")
    val appIcon = requireArguments().getInt(ARG_APP_ICON, 0)
    val appIsSystem = requireArguments().getBoolean(ARG_APP_SYSTEM, false)
    require(appPackageName.isNotBlank())
    require(appName.isNotBlank())
    val listStateTag = TAG + appPackageName

    Injector.obtain<PadLockComponent>(requireContext().applicationContext)
        .plusLockInfoComponent()
        .activity(requireActivity())
        .owner(viewLifecycleOwner)
        .appName(appName)
        .packageName(appPackageName)
        .appIcon(appIcon)
        .appSystem(appIsSystem)
        .listStateTag(listStateTag)
        .inflater(inflater)
        .container(container)
        .savedInstanceState(savedInstanceState)
        .diffProvider(object : ListDiffProvider<ActivityEntry> {
          override fun data(): List<ActivityEntry> = lockView.getListData()
        })
        .build()
        .inject(this)

    lockView.create()
    return lockView.root()
  }

  override fun onViewCreated(
    view: View,
    savedInstanceState: Bundle?
  ) {
    super.onViewCreated(view, savedInstanceState)

    lockView.onSwipeRefresh { populateList(true) }

    lockView.onToolbarNavigationClicked { dismiss() }

    lockView.onToolbarMenuItemClicked {
      if (it == R.id.menu_explain_lock_type) {
        LockInfoExplanationDialog().show(requireActivity(), "lock_info_explain")
      }
    }

    databaseChangeDisposable = viewModel.onDatabaseChangeEvent(
        onChange = { lockView.onDatabaseChangeReceived(it.index, it.entry) },
        onError = { lockView.onDatabaseChangeError { populateList(true) } }
    )

    lockEventDisposable = viewModel.onLockEvent(
        onWhitelist = { populateList(true) },
        onError = { lockView.onModifyEntryError { populateList(true) } }
    )
  }

  private fun populateList(forced: Boolean) {
    populateDisposable = viewModel.populateList(forced,
        onPopulateBegin = { lockView.onListPopulateBegin() },
        onPopulateSuccess = { lockView.onListLoaded(it) },
        onPopulateError = { lockView.onListPopulateError { populateList(true) } },
        onPopulateComplete = { lockView.onListPopulated() }
    )
  }

  override fun onDestroyView() {
    super.onDestroyView()

    databaseChangeDisposable.tryDispose()
    lockEventDisposable.tryDispose()
    populateDisposable.tryDispose()
  }

  override fun onStart() {
    super.onStart()
    populateList(false)
  }

  override fun onPause() {
    super.onPause()
    lockView.commitListState(null)
  }

  override fun onSaveInstanceState(outState: Bundle) {
    lockView.commitListState(outState)
    super.onSaveInstanceState(outState)
  }

  override fun onResume() {
    super.onResume()
    // The dialog is super small for some reason. We have to set the size manually, in onResume
    dialog.window?.apply {
      setLayout(
          WindowManager.LayoutParams.MATCH_PARENT,
          WindowManager.LayoutParams.WRAP_CONTENT
      )
      setGravity(Gravity.CENTER)
    }
  }

  companion object {

    internal const val TAG = "LockInfoDialog"
    private const val ARG_APP_PACKAGE_NAME = "app_packagename"
    private const val ARG_APP_ICON = "app_icon"
    private const val ARG_APP_NAME = "app_name"
    private const val ARG_APP_SYSTEM = "app_system"

    @CheckResult
    @JvmStatic
    fun newInstance(appEntry: AppEntry): LockInfoDialog {
      return LockInfoDialog().apply {
        arguments = Bundle().apply {
          putString(ARG_APP_PACKAGE_NAME, appEntry.packageName)
          putString(ARG_APP_NAME, appEntry.name)
          putInt(ARG_APP_ICON, appEntry.icon)
          putBoolean(ARG_APP_SYSTEM, appEntry.system)
        }
      }
    }
  }
}
